docs/csharp/local-functions-vs-lambdas.md

```diff
@@ -26,46 +27,88 @@ Contrast that implementation with a version that uses lambda expressions:
+The local functions have names. The lambda expressions are anonymous methods
+that are assigned to variables that are `Func` or `Action` types. When you
+declare a local function, the argument types and return type are part of the 
+function declaration. Instead of being part of the body of the lambda expression, the argument types and return type are part of the lambda
+expression's variable type declaration. Those two differences may
+result in clearer code.
+
+Local functions have different rules for definite assignment
+than lambda expressions. A local function declaration can be referenced
+from any code location where it is in scope. A lambda expression must be
+assigned before it can be accessed (or called through a delgate
+referencing the lambda expression.) Notice that the version using the
+lambda expression must declare and initialize the lambda expression,
+`nthFactorial` before defining it. Not doing so results in a compile
+time error for referencing `nthFactorial` before assigning it.
+These differences mean that recursive algorithms are easier to create
+using local functions. You can declare and define a local function that
+calls itself. Lambda expressions must be declared, and assigned a default
+value before they can be re-assigned to a body that references the same
+lambda expression.
+
+Definite assignment rules also affect any variables that are captured
+by the local function or lamdba epression. Both local functions and
+lambda expression rules demand that any captured variables are definitely
+assigned at the point when the local function or lambda expression is
+converted to a delegate. The difference is that lambda expressions are converted
+to delegates when they are declared. Local functions are converted to delegates
+only when used as a delegate. If you declare a local function and only
+reference it by calling it like a method, it will not be converted to
+a delegate. That rule enables you to declare
+a local function at any convenient location in its enclosing scope. It's common
+to declare local functions at the end of the parent method, after any return
+statements.
+
+Third, the compiler can perform static analysis that enables local functions to
+definitely assign captured variables in the enclosing scope. Consider this example:
+
+```csharp
+bool M()
+{
+    int y;
+    Local();
+    return y;
+
+    void Local() => y = 0;
+}
+```
+
+The compiler can determine that `Local` definitely assigns `y` when called. Because `Local` is called before the `return` statement, `y` is definitiely
+assigned at the `return` statement.
+
+The analysis that enables that analysis enables the fourth difference.
+Depending on their use, local functions can avoid heap allocations that
+are always necessary for lambda expressions. If a local function is never
+converted to a delegate and none of the variables captured by a locatl function
+are converted to delegates, the compiler can avoid heap allocations. 
```

Comment:
Technically, it should read "none of the variables captured by the local function are captured by other lambdas or local functions that are converted to delegates." That is, the variables can't be captured by anything converted to a delegate, even if that capturing is done in another closure.